### PR TITLE
Remove parts in order for object storage without zero-copy

### PR DESCRIPTION
**Pull request by vdimir**
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
* Fix error when some blobs from s3 without zero-copy replication for outdated part are not removed
 
**Questions:**
- [ ] Should we do the same for regular MergeTree over object storage (not only for Replicated)?

cc @KochetovNicolai @tavplubix

### CODE REVIEW

**Comment by robot-ch-test-poll1** [comment]
This is an automated comment for commit 895e948fa9352d8cd7f8930eaeaa965779c29ef1 with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/51900/895e948fa9352d8cd7f8930eaeaa965779c29ef1/ci_running.html)
The overall status of the commit is 🟡 pending

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🟡 pending</td></tr>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>🟢 success</td></tr>
</table>

File: src/Storages/MergeTree/MergeTreeData.cpp [hunk 1]
```diff
@@ -2137,20 +2137,18 @@ MergeTreeData::DataPartsVector MergeTreeData::grabOldParts(bool force)
...
+    bool need_remove_parts_in_order = false;
 
-    if (need_remove_parts_in_order)
+    if (supportsReplication())
     {
-        bool has_zero_copy_disk = false;
         for (const auto & disk : getDisks())
         {
-            if (disk->supportZeroCopyReplication())
+            if (disk->supportZeroCopyReplication() || disk->isRemote())
```
Comment (KochetovNicolai) [inline src/Storages/MergeTree/MergeTreeData.cpp:2146]: I think `disk->supportZeroCopyReplication() => disk->isRemote()`

File: src/Storages/MergeTree/MergeTreeData.cpp [hunk 2]
```diff
@@ -2397,15 +2395,18 @@ void MergeTreeData::clearPartsFromFilesystemImpl(const DataPartsVector & parts_t
...
     bool remove_parts_in_order = false;
-    if (settings->allow_remote_fs_zero_copy_replication && dynamic_cast<StorageReplicatedMergeTree *>(this) != nullptr)
+    if (dynamic_cast<StorageReplicatedMergeTree *>(this) != nullptr)
     {
         remove_parts_in_order = std::any_of(
             parts_to_remove.begin(), parts_to_remove.end(),
-            [] (const auto & data_part) { return data_part->isStoredOnRemoteDiskWithZeroCopySupport() && data_part->info.getMutationVersion() > 0; }
+            [] (const auto & data_part)
+            {
+                return (data_part->isStoredOnRemoteDiskWithZeroCopySupport() || data_part->isStoredOnRemoteDisk())
```
Comment (KochetovNicolai) [inline src/Storages/MergeTree/MergeTreeData.cpp:2404]: And `data_part->isStoredOnRemoteDiskWithZeroCopySupport()=>| data_part->isStoredOnRemoteDisk()`

**Comment by tavplubix** [comment]
Why do we need to remove parts in order without zero-copy replication? Please add comments

**Comment by vdimir** [comment]
> Why do we need to remove parts in order without zero-copy replication? Please add comments

@tavplubix 

I've added comments. I believe there can be better fix for this case, since there's some race condition on blobs removal when we remove parts that have hard-links. We can try to fix it directly and leave concurrent parts removal as it is.
However, for now, maybe we can go with the same approach as we do for zero copy, because we want to have a fix simple enough to back-port, in which we can rely on the existing code instead of changes in other places.

**Comment by tavplubix** [comment]
> some race condition on blobs removal 

What race condition? 

> parts that have hard-links

With zero-copy replication, only files in mutated parts may have hardlink. Hardlinks in other places are disabled for zero-copy replication. But without zero-copy, we have much more places where hardlinks are used. It makes sense to fix them or disable.

**Comment by tavplubix** [comment]
Closed in favor of https://github.com/ClickHouse/ClickHouse/pull/51954

**Pull request closed without merging**